### PR TITLE
Redistribute metadata kind constants to delineate ABI vs private values.

### PR DESCRIPTION
**Pull request by jckarter**
And more cleanly separate type and heap while we're at it, making it a bit easier to allocate new compiler-generated metadata kinds without worrying about stepping on the toes of runtime-generated kinds, and ask basic questions like "is this type and/or heap metadata" in a forward- and backward-compatible way.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

File: include/swift/ABI/MetadataValues.h [hunk 1]
```diff
@@ -47,6 +47,21 @@ struct InProcess;
...
 using Metadata = TargetMetadata<InProcess>;
 
+/// Non-type metadata kinds have this bit set.
+const unsigned MetadataKindIsNonType = 0x400;
+
+/// Non-heap metadata kinds have this bit set.
+const unsigned MetadataKindIsNonHeap = 0x200;
+
+// The above two flags are negative because the "class" kind has to be zero,
+// and class metadata is both type and heap metadata.
```
Comment (rjmccall) [inline include/swift/ABI/MetadataValues.h]: Why does the class kind have to be zero?  It's not like this scheme allows us to efficiently work with ObjC-compatible classes.

Comment (jckarter) [reply include/swift/ABI/MetadataValues.h]: I was thinking specifically about how ObjC interop requires us to put the isa field where the kind normally goes, so we have to treat a value >= 4096 as meaning `kind == Class`, but you're right that that doesn't *require* `Class` to be zero specifically.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test

**Pull request merged**